llvm/lib/Transforms/Utils/LoopMetadata.cpp

```diff
@@ -0,0 +1,235 @@
+  assert(MinIterCount > 0);
+
+  // extract loop counter increment/decrement
+  const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(S);
+  if (!AR) {
+    LLVM_DEBUG(dbgs() << "LoopMetadata-Warning: SCEV S i not a SCEVAddRecExpr ";
+               S->dump());
+    return nullptr;
+  }
+  const SCEV *ConstExpr = AR->getOperand(1);
+  if (!AR->isAffine() || !isa<SCEVConstant>(ConstExpr)) {
+    LLVM_DEBUG(dbgs() << "LoopMetadata-Warning: Unknown SCEVAddRecExpr ";
+               AR->dump());
+    return nullptr;
+  }
+
+  int MinIterValue = std::abs(LoopStepSize * MinIterCount);
+  // calculate the minimum iteration value, since SGE is used, subtract 1
+  MinIterValue--;
+
+  return llvm::ConstantInt::get(llvm::Type::getInt32Ty(*Context), MinIterValue,
+                                true);
+}
+
+Value *getLoopInvariant(Value *Op0, Value *Op1, ScalarEvolution *SE) {
+  if ((SE->isSCEVable(Op0->getType()) &&
+       (SCEVAddExpr::classof(SE->getSCEV(Op0)) ||
+        SCEVAddRecExpr::classof(SE->getSCEV(Op0)))) ||
+      isa<PHINode>(Op0))
+    return Op1;
+
+  return Op0;
+}
+
+void LoopMetadata::getBoundaries(const SCEV *S) {
+  assert(SCEVAddRecExpr::classof(S));
+
+  Value *Op0 = LoopCmpInstr->getOperand(0);
+  Value *Op1 = LoopCmpInstr->getOperand(1);
+
+  const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(S);
+  const SCEVConstant *SCEVConst = dyn_cast<SCEVConstant>(AR->getStart());
+
+  if (IsLoopIncrementing) {
+    if (!SCEVConst) {
+      LowerBoundary = nullptr;
+      UpperBoundary = nullptr;
+      return;
+    }
+    LowerBoundary = SCEVConst->getValue();
+    UpperBoundary = getLoopInvariant(Op0, Op1, SE);
+
+  } else {
+
+    if (SCEVConst)
+      UpperBoundary = SCEVConst->getValue();
+
+    if (const SCEVUnknown *S = dyn_cast<SCEVUnknown>(AR->getStart()))
+      UpperBoundary = S->getValue();
+
+    LowerBoundary = getLoopInvariant(Op0, Op1, SE);
+  }
+
+}
+
+const SCEV *LoopMetadata::getAddRecSCEV(Value *Op) {
+  if (!SE->isSCEVable(Op->getType()))
+    return nullptr;
+
+  const SCEV *S = SE->getSCEV(Op);
+  if (SCEVAddRecExpr::classof(S))
+    return S;
+
+  return nullptr;
+}
+
+const SCEV *LoopMetadata::getSCEV() {
+  for (Value *Op : LoopCmpInstr->operands()) {
+    if (const SCEV *S = getAddRecSCEV(Op))
+      return S;
+  }
+
+  return nullptr;
+}
+
+
+int getSCEVStepSize(const SCEV *S) {
+  assert(isa<SCEVConstant>(S));
+  return dyn_cast<SCEVConstant>(S)->getValue()->getSExtValue();
+}
+
+void LoopMetadata::addAssumeToLoopHeader() {
+  // reset Loop specific information;
+  LowerBoundary = nullptr;
+  UpperBoundary = nullptr;
+
+  LLVM_DEBUG(dbgs() << "Processing Loop Metadata: "
+                    << L->getHeader()->getParent()->getName() << " "
+                    << L->getName() << " (" << MinIterCount << ")\n");
+
+  BranchInst *BI = dyn_cast<BranchInst>(L->getExitingBlock()->getTerminator());
+  if (!BI)
+    return;
+
+  LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return;
```

Comment:
I think we assume here that the terminator is a conditional branch. We also never check the relation of the compare instruction, which indicates an over-optimistic view on a hostile world.